kubespawner/proxy.py

```diff
@@ -97,23 +96,45 @@ def _namespace_default(self):
         config=True,
         help="""
         Full host name of the k8s API server ("https://hostname:port").
-        
-        Typically this is unnecessary, the hostname is picked up by 
+
+        Typically this is unnecessary, the hostname is picked up by
         config.load_incluster_config() or config.load_kube_config.
         """,
     )
 
     def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
 
-        # We use the maximum number of concurrent user server starts (and thus proxy adds)
-        # as our threadpool maximum. This ensures that contention here does not become
-        # an accidental bottleneck. Since we serialize our create operations, we only
-        # need 1x concurrent_spawn_limit, not 3x.
-        self.executor = ThreadPoolExecutor(max_workers=self.app.concurrent_spawn_limit)
+        # Schedules async initialization logic that is to be awaited by async
+        # functions by decorating them with @_await_async_init.
+        self._async_init_future = asyncio.ensure_future(self._async_init())
 
-        # Global configuration before reflector.py code runs
-        self._set_k8s_client_configuration()
+    async def _async_init(self):
+        """
+        This method is scheduled to run from `__init__`, but not awaited there
+        as `__init__` can't be marked as async.
+
+        Since JupyterHub won't await this method, we ensure the async methods
+        JupyterHub may call on this object will await this method before
+        continuing. To do this, we decorate them with `_await_async_init`.
+
+        But, how do we figure out the methods to decorate? Likely only those
+        exposed by the base class that JupyterHub would know about. The base
+        class is Proxy, as declared in proxy.py:
+        https://github.com/jupyterhub/jupyterhub/blob/HEAD/jupyterhub/proxy.py.
+
+        From the Proxy class docstring we can conclude that the following
+        methods, if implemented, could be what we need to decorate with
+        _await_async_init:
+
+          - get_all_routes (implemented and decorated)
```

Comment:
Thanks Yuvi for reviewing this!

This is a sound concern but I'm failing to address it with the suggestion as a blob of complexity makes it hard for me to even describe why: but I'll try below.

`shared_client` is now used from `_async_init` in KubeSpawner (where `self.api` is set) and KubeIngressProxy (were `self.core_api` and `self.extension_api` is set). `load_config` is awaited just before calling `shared_client`, so from that perspective it's not an issue of making `shared_client` a async getter, but doing so would not resolve the situation. We still need `_async_init` to finish for various other methods, because they reference `self.api` / `self.core_api` - so we need to ensure those won't run before we have initialized `self.api`.

# Ideas considered

- __Idea 1__
  To ensure that when we need to use an api client, we don't rely on something set on the instance like `self.api`, but request it fresh from `shared_client`.
  - Ugly detail: `k8s_api_host` traitlet's value for example is needed by `load_config`, which now is to be part of `shared_client`'s logic. This traitlet's value is not available to the reflector classes even though they are part of KubeSpawner/KubeIngressProxy. This is likely addressable, but I find it to be a bit tricky on passing the relevant information onwards to the reflector and I'm not sure how to do it in a relatively clean way.
  - Ugly detail: requesting the api client on demand, also now `async`, means we need to await the client first. Following that we also await some some function on it. This is not a technical problem, just bloats the code a bit.
- __Idea 2__
  We ensure that whenever we want to use `self.api` etc from a function we have defined, we do it from a function decorated with `_await_async_init`. That way, we have control within this project when it may be needed. This idea benefits from the overview below about where things are used.
- __Idea 3__
  We double down on the current solution, where we have relied on understanding JupyterHub's Spawner/Proxy base classes which is the interface JupyterHub use to work against our code in this project. While we rely on JupyterHub's internal working, we rely on the public interface we expose to JupyterHub by being a Spawner class (and Proxy class).
  
  I think this could lead to some documentation improvements in JupyterHub. I believe an overview of all interactions JupyterHub has with the Spawner / Proxy class can be improved to for example cover all hooks and functions in a single page for situations like this.

# Where are things used?

## `shared_client` is used from...

spawner.py:
- `_async_init`

proxy.py:
- `_async_init`

reflector.py:
- `__init__` (`load_config` is assumed to have run here)

## The api clients are used from...

`self.api` in spawner.py is used from these functions:
- `_async_init` (initialized)
- `_make_create_pod_request`
- `_make_create_pvc_request`
- `_ensure_not_exists`
- `_make_create_resource_request`
- `_make_delete_pod_request`
- `_make_delete_pvc_request`
- `_ensure_namespace`

`self.core_api` and `self.extension_api` in proxy.py is used from these functions:
- `_async_init` (initialized)
- `add_route`
- `delete_route`

`self.api` in reflector.py is used from these functions:
- `__init__` (initialized)
- `_list_and_update`
- `_watch_and_update`

Reply:
Yes, as @consideRatio ratio said, there are two loosely-coupled problems here.

1) `shared_client` needs to know the traitlet values in order to mess with the TLS settings
2) The client has to be configured before use, but we don't know what method will be first used and we don't know that it even will be one of the documented methods rather than something happening in a hook.

It feels like there's something we can do with making a SharedClient mixin class that the Spawner and Proxy both inherit...

...but I think maybe the thing to do is go with what we have now and put that into the whole bundle of investigation about how much TLS setup really costs.  Conceptually it would be a lot cleaner to NOT share the client and treat it as an async context manager, which is how the underlying library wants you to use it.  I did not go with this approach because while I'm sure it's going to look fine in the test cases, I really want to try that out at n=500 or so in order to see whether it scales.

Reply:
...and even if it's a mixin there's still the problem that sooner or later you're going to have to `await()` the configuration one way or another, and thus that can't happen from `__init__()`.  I think I am in favor of changing the calling conventions from JupyterHub to do something like call a `create()` classmethod, if such exists, instead of just instantiating a new instance of the class, but I don't know how feasible asking for that upstream would be.

Reply:
@minrk would it be reasonable to make JupyterHub look for if the Spawner / Proxy class provides a async constructor, and if so, make JupyterHub await it?

Reply:
Specifically, one with a well-known name (I like `create()`).  The logic would look like:

for a thing with a specified class (e.g. Spawner, Proxy, dunno if there's anything else):

Does that class have an async  classmethod named `create()`?  If so, the instance of the class is created with `await kls.create()`; if not it's just `kls()` (the current behavior)

Reply:
I inspected how JupyterHub interacted with an initialized Spawner object, and it seems its quite hard to overview. I think the right call is to go with Idea 2 - to decorate all the methods in KubeSpawner we know will end up needing to await the `_async_init` call that initialized the `shared_client` ahead of time before it is used.

## JupyterHub interactions with Spawner class instances

The [`_new_spawner` function](https://github.com/jupyterhub/jupyterhub/blob/a3ea0f0449e9e19feff863b5f5cce4d8a4febf1f/jupyterhub/user.py#L379) is used as a (sync) factory function.

It is used only once, [in the (sync) `__init__ function](https://github.com/jupyterhub/jupyterhub/blob/a3ea0f0449e9e19feff863b5f5cce4d8a4febf1f/jupyterhub/user.py#L242) of the User class, which has a [`_SpawnerDict` class instance](https://github.com/jupyterhub/jupyterhub/blob/a3ea0f0449e9e19feff863b5f5cce4d8a4febf1f/jupyterhub/user.py#L210-L217) create a Spawner class whenever the `_SpawnerDict` instance, now available in the user instance `spawners` property is referenced. For example, via `my_user.spawners["any_key_here"]` - then a Spawner instance is returned in a sync manner.

Since JupyterHub inits Spawner instances from sync functions, and sync properties indirectly, it seems like a relatively major change to rely on a async factory function.

## Observed functions used by User object

- [`_delete_spawner` function](https://github.com/jupyterhub/jupyterhub/blob/a3ea0f0449e9e19feff863b5f5cce4d8a4febf1f/jupyterhub/user.py#L344) calls:
  - `delete_forever()`
- [`running` property](https://github.com/jupyterhub/jupyterhub/blob/a3ea0f0449e9e19feff863b5f5cce4d8a4febf1f/jupyterhub/user.py#L457-L462) could init a Spawner object
- [`server` property](https://github.com/jupyterhub/jupyterhub/blob/a3ea0f0449e9e19feff863b5f5cce4d8a4febf1f/jupyterhub/user.py#L487-L489) could init a Spawner object
- [`spawn` function](https://github.com/jupyterhub/jupyterhub/blob/a3ea0f0449e9e19feff863b5f5cce4d8a4febf1f/jupyterhub/user.py#L605) calls:
  - `clear_state()`
  - `pre_spawn_start()`
  - `run_auth_state_hook()`
  - `run_pre_spawn_hook()`
  - `create_certs()`
  - `move_certs()`
  - `start()`
  - `start_polling()`
- [`stop` function](https://github.com/jupyterhub/jupyterhub/blob/a3ea0f0449e9e19feff863b5f5cce4d8a4febf1f/jupyterhub/user.py#L885) calls:
  - `stop_polling()`
  - `poll()`
  - `stop()`
  - `run_post_stop_hook()`
  - `clear_state()`
  - `get_state()`

Reply:
@yuvipanda @athornton I've implemented `Idea 2` in https://github.com/lsst-sqre/kubespawner/pull/13, a PR to this PR.

Reply:
@minrk @yuvipanda @athornton I'm really struggling to get somewhere. I'm running low on steam =/ I've run into test failures in https://github.com/lsst-sqre/kubespawner/pull/13.

The clean way of ensuring we await the async init is simply to do it from JupyterHub with a custom async constructor I think... I looked into the JupyterHub code base about that recently, and it seems quite tricky as well. See https://github.com/jupyterhub/jupyterhub/pull/3801 PR for comments about it.